src/main/java/picard/sam/SplitSamBySize.java

```diff
@@ -0,0 +1,162 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam;
+
+import htsjdk.samtools.*;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.ProgressLogger;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineProgramProperties;
+import org.broadinstitute.barclay.help.DocumentedFeature;
+import picard.cmdline.CommandLineProgram;
+import picard.cmdline.StandardOptionDefinitions;
+import picard.cmdline.programgroups.SamOrBam;
+
+import java.io.File;
+import java.text.DecimalFormat;
+
+/**
+ * <p/>
+ * Splits the input queryname sorted or query-grouped SAM/BAM file and writes it into
+ * multiple BAM files, each with an approximately equal number of reads. This will retain the sort order
+ * within each output BAM and if the BAMs are concatenated in order (output files are named
+ * numerically) the order of the reads will match the original BAM.
+ */
+@CommandLineProgramProperties(
+        summary = SplitSamBySize.USAGE_SUMMARY + SplitSamBySize.USAGE_DETAILS,
+        oneLineSummary = SplitSamBySize.USAGE_SUMMARY,
+        programGroup = SamOrBam.class)
+@DocumentedFeature
+public class SplitSamBySize extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Splits a SAM or BAM file to multiple BAMs.";
+    static final String USAGE_DETAILS = "This tool splits the input query-grouped SAM/BAM file into multiple BAM files " +
+            "while maintaining the sort order. This can be used to split a large unmapped BAM in order to parallelize alignment."+
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SplitSamBySize \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000 \\ <br />" +
+            "     SPLIT_TO_N_READS=48000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N_READS", doc = "Split to have approximately N reads per output file.", mutex = {"SPLIT_TO_N_FILES"})
+    public int SPLIT_TO_N_READS;
+
+    @Argument(shortName = "N_FILES", doc = "Split to N files.", mutex = {"SPLIT_TO_N_READS"})
+    public int SPLIT_TO_N_FILES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file.")
+    public int TOTAL_READS_IN_INPUT;
```

Comment:
I am not sure why this option is needed when `SPLIT_TO_N_READS` is given.  Can't we just output the first `SPLIT_TO_N_READS` reads, then create a new writer?

Reply:
The problem I'm trying to avoid when I don't have `TOTAL_READS_IN_INPUT` is that there could be a remainder number of reads in the final file that ranges from none to `SPLIT_TO_N_READS`. Since the use case I was thinking of was to split for alignment, I want to avoid aligning a file with too small a number of reads in it. One potential solution to this would be to combine the last two files, but then you run the risk of having one file that's double the size of the others, which also isn't ideal.

Reply:
I see, I just think knowing the total # of input reads ahead of time is onerous or heavy-handed for users.   Why not just have re-write the last two files if the last one will have too few reads?

Reply:
Do you mean, to only combine the last two files if the last file has two few reads and otherwise leave them as two files? One potential issue there is that I have to define what is too few reads. I suppose I could make that a user specified value? 

Another option could be to have this tool optionally calculate the `TOTAL_READS_IN_INPUT` itself, but then wouldn't I either have to traverse the BAM twice or have an index file? Is there a quick way to get the total number of reads in Picard?

Reply:
Ok, I think I have a new solution. Instead of splitting the file to keep number of reads approximately equivalent in each file, I can split by on disk file size so that the number of bytes in the output files will be approximately equivalent. Obviously I'll still keep query-grouped reads together. This eliminates needing to know the number of reads in the input since it's very easy to get the size of the file.